cmd/tree-main.go

```diff
@@ -0,0 +1,227 @@
+      $ {{.HelpName}} s3/mybucket/
+
+   3. List all contents of mybucket on Amazon S3 cloud storage on Microsoft Windows in a tree format.
+      $ {{.HelpName}} s3\mybucket\
+`,
+}
+
+// checkTreeSyntax - validate all the passed arguments
+func checkTreeSyntax(ctx *cli.Context) {
+	args := ctx.Args()
+	if !ctx.Args().Present() {
+		args = []string{"."}
+	}
+	for _, arg := range args {
+		if strings.TrimSpace(arg) == "" {
+			fatalIf(errInvalidArgument().Trace(args...), "Unable to validate empty argument.")
+		}
+	}
+	// extract URLs.
+	URLs := ctx.Args()
+
+	for _, url := range URLs {
+		_, _, err := url2Stat(url, false, nil)
+		if err != nil && !isURLPrefixExists(url, false) {
+			// Bucket name empty is a valid error for 'ls myminio',
+			// treat it as such.
+			if _, ok := err.ToGoError().(BucketNameEmpty); ok {
+				continue
+			}
+			fatalIf(err.Trace(url), "Unable to stat `"+url+"`.")
+		}
+	}
+}
+
+// doTree - list all entities inside a folder in a tree format.
+func doTree(url string, level int, leaf bool, dirClosed map[int]bool) error {
+
+	targetAlias, targetURL, _ := mustExpandAlias(url)
+	if !strings.HasSuffix(targetURL, "/") {
+		targetURL += "/"
+	}
+
+	clnt, err := newClientFromAlias(targetAlias, targetURL)
+	fatalIf(err.Trace(targetURL), "Unable to initialize target `"+targetURL+"`.")
+
+	prefixPath := clnt.GetURL().Path
+	separator := string(clnt.GetURL().Separator)
+	if !strings.HasSuffix(prefixPath, separator) {
+		prefixPath = prefixPath[:strings.LastIndex(prefixPath, separator)+1]
+	}
+
+	bucketNameShowed := false
+	var prev *clientContent
+	show := func(end bool) error {
+		if level == 0 && !bucketNameShowed {
+			bucketNameShowed = true
+			fmt.Printf("%s\n", console.Colorize("Dir", url))
+		}
+
+		var branchString string
+		if level != 0 {
+			for i := 0; i < level; i++ {
+				if dirClosed[i] {
+					branchString += treeLevel
+				} else {
+					branchString += treeNext + treeLevel
+				}
+			}
+		}
+
+		if end {
+			dirClosed[level] = true
+			branchString += treeLastEntry
+		} else {
+			dirClosed[level] = false
+			branchString += treeEntry
+		}
+
+		// Convert any os specific delimiters to "/".
+		contentURL := filepath.ToSlash(prev.URL.Path)
+		prefixPath = filepath.ToSlash(prefixPath)
+
+		// Trim prefix of current working dir
+		prefixPath = strings.TrimPrefix(prefixPath, "."+separator)
+
+		var entry string
+		if prev.Type.IsDir() {
+			entry = console.Colorize("Dir", strings.TrimPrefix(contentURL, prefixPath))
+		} else {
+			entry = console.Colorize("File", strings.TrimPrefix(contentURL, prefixPath))
+		}
+
+		fmt.Printf("%s%s\n", branchString, entry)
+
+		if prev.Type.IsDir() {
+			url := ""
+			if targetAlias != "" {
+				url = targetAlias + "/" + contentURL
+			} else {
+				url = contentURL
+			}
+
+			if err := doTree(url, level+1, end, dirClosed); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	for content := range clnt.List(false, false, DirNone) {
```

Comment:
There should be a bug in `Client.List()`, For `ListAllMyBuckets` S3 calls, it should return each bucket entry as directory is `Type.IsDir()` should return true, but its not. else S3 level tree works normally.  Please confirm with this @abperiasamy and fix it accordingly.